app/javascript/shopify_app/turbolinks.js

```diff
@@ -0,0 +1,58 @@
+import { getSessionToken } from "@shopify/app-bridge-utils";
+
+const SESSION_TOKEN_REFRESH_INTERVAL = 2000; // Request a new token every 2s
+
+document.addEventListener("turbolinks:request-start", function (event) {
+  var xhr = event.data.xhr;
+  xhr.setRequestHeader("Authorization", "Bearer " + window.sessionToken);
+});
+
+document.addEventListener("turbolinks:render", function () {
+  $("form, a[data-method=delete]").on("ajax:beforeSend", function (event) {
+    const xhr = event.detail[0];
+    xhr.setRequestHeader("Authorization", "Bearer " + window.sessionToken);
+  });
+});
+
+document.addEventListener("DOMContentLoaded", async function () {
+  // Wait for a session token before trying to load an authenticated page
+  await retrieveToken(app);
+
+  // Keep retrieving a session token periodically
+  keepRetrievingToken(app);
```

Comment:
I'm guessing there is absolutely no way to subscribe to a session token change event to avoid polling?

Reply:
> absolutely no way

It would _seem_ that way! 😅 I'm still exploring options that don't require us to poll (this would be ideal).

Reply:
I was looking at Basecamp's new replacement for Turbolinks: Turbo Drive. https://github.com/hotwired/turbo/tree/main

It's also not built async, but it's in typescript and is in active development. We could consider trying to switch it to be mostly async, and then we'd be able to `await` in the `before-visit` hook. That's a big project though I think.

Reply:
I built a request interceptor for Turbo Drive. It allows inserting custom function before every Turbo request and form submission. Here's the interceptor code: https://github.com/kirillplatonov/turbo/commit/b58cfcedd2aada39995ab9fb6a27387ea2e5070a

And that's how `shopify_app.js` looks with it:
```javascript
import { Turbo } from "@hotwired/turbo-rails";
import { getSessionToken } from "@shopify/app-bridge-utils";

document.addEventListener('DOMContentLoaded', () => {
  var data = document.getElementById('shopify-app-init').dataset;
  var AppBridge = window['app-bridge'];
  var createApp = AppBridge.default;
  window.app = createApp({
    apiKey: data.apiKey,
    shopOrigin: data.shopOrigin,
  });

  var actions = AppBridge.actions;
  var TitleBar = actions.TitleBar;
  TitleBar.create(app, {
    title: data.page,
  });

  // Redirect to the requested page
  Turbo.visit(data.loadPath);
});

// Intercept every Turbo request and load Shopify session token
Turbo.setRequestInterceptor(async (request) => {
  const token = await getSessionToken(window.app);
  request.addHeader("Authorization", `Bearer ${token}`);
});

// Fix redirect_to issue in Safari (it looses custom headers during 302 redirect)
document.addEventListener("turbo:before-fetch-response", (event) => {
  const response = event.detail.fetchResponse;
  const status = response.statusCode;
  const location = response.header("Location");

  if (status === 300 && location !== null) {
    event.preventDefault();
    Turbo.visit(location);
  }
});
```

No more polling 🥳